text/0000-module-resolution.md

```diff
@@ -0,0 +1,340 @@
+
+Here is how the structure of the repo looks like  you can take a look at [ a real implementation implementation and test here](https://github.com/salesforce/lwc/pull/1602/files#diff-5cfc908f1b15979cf59306304092ddb5):
+
+```markup
+rootDir
+├── src/
+│   └── modules/
+│       └── root/
+│           └── foo/
+│               └── foo.js ("root/foo")
+└── node_modules/
+    ├── @ui/
+    │   └── components/
+    │       ├── package.json [ui-components@2.0.0]
+    │       └── src/
+    │           └── modules/
+    │               └── ui/
+    │                   └── button/
+    │                       └── button.js ("ui/button")
+    └── fancy-components/
+        ├── package.json     [fancy-components@1.0.0]
+        ├── src/
+        │   └── modules/
+        │       └── fancy/
+        │           └── bar/
+        │               └── bar.js ("fancy/bar")
+        └── node_modules/
+            └── @ui/
+                └── components/
+                    ├── package.json [ui-components@v1.0.0]
+                    └── src/
+                        └── modules/
+                            └── ui/
+                                └── button/
+                                    └── button.js ("ui/button")
+```
+
+When importing those components from a given module here is what we would expect:
+
+```js
+// importer: root/foo/foo.js
+import button from "ui/button"; // should resolve against ui/button@2.0.0
+```
+
+```js
+// importer: fancy/bar/bar.js
+import button from "ui/button"; // should resolve against ui/button@1.0.0
+```
+
+This means that from a developers perspective you just import the module you need.
+
+## State of the art
+
+An important topic to understand first is how module resolution works today in different technologies and host environments.
+
+JavaScript did not had built-in support for modules, but the community has created impressive work-arounds over the years. The two most important (and unfortunately incompatible) standards are: CommonJS (used fundamentally by Node - with some extra semantics) and AMD (Asyncronous module definition), used by RequireJs and a lot of other bundlers.
+
+Note that there are other non-standard formats, loaders and tools like UMD, SystemJS, IIFE, etc, which are interesting for you know and understand but are out of the scope for this proposal.
+
+### Node.js resolution
+
+If you ever used Node, you know that in order to import a module you need to use the `require()` method, and if you want to export something to some other module or file you must use `module.exports`. That is know nas [CommonJS format](https://en.wikipedia.org/wiki/CommonJS). Now if you want to import a 3party package how does that works? Well Node decided to implement its own resolution by looking at the content of a the node_modules folder based on some dependencies defined on `package.json`. You can look at the full algorithm [here](https://nodejs.org/docs/latest/api/modules.html#modules_all_together).
```

Comment:
If you ever used Node, you know that in order to import a module you need to use the `require()` method, and if you want to export something to some other module or file you must use `module.exports`. That is known as [CommonJS format](https://en.wikipedia.org/wiki/CommonJS). Now if you want to import a 3party package how does that work? Well Node decided to implement its own resolution by looking at the content of the `node_modules` folder based on some dependencies defined on `package.json`. You can look at the full algorithm [here](https://nodejs.org/docs/latest/api/modules.html#modules_all_together).